Tests/ApolloCodegenTests/ApolloCodegenTests.swift

```diff
@@ -0,0 +1,193 @@
+import XCTest
+import ApolloCodegenTestSupport
+@testable import ApolloCodegenLib
+import Nimble
+
+class ApolloCodegenTests: XCTestCase {
+  private let schemaSDL: String = {
+    """
+    type Query {
+      books: [Book!]!
+      authors: [Author!]!
+    }
+
+    type Book {
+      title: String!
+      author: Author!
+    }
+
+    type Author {
+      name: String!
+      books: [Book!]!
+    }
+    """
+  }()
+
+  let directoryURL = CodegenTestHelper.outputFolderURL().appendingPathComponent("Codegen")
+
+  override func setUpWithError() throws {
+    try FileManager.default.apollo.createDirectoryIfNeeded(atPath: directoryURL.path)
+  }
+
+  override func tearDownWithError() throws {
+    try cleanTestOutput()
+  }
+
+  // MARK: Helpers
+
+  private func cleanTestOutput() throws {
+    try FileManager.default.apollo.deleteDirectory(atPath: directoryURL.path)
+  }
+
+  private func createSchema(at path: String) {
+    expect(
+      try FileManager.default.apollo.createFile(
+        atPath: path,
+        data: self.schemaSDL.data(using: .utf8)!
+      )
+    ).notTo(throwError())
+  }
+
+  private func createOperationFile(for data: Data, inDirectory directoryURL: URL) {
+    expect(
+      try FileManager.default.apollo.createFile(
+        atPath: directoryURL.appendingPathComponent("\(UUID().uuidString).graphql").path,
+        data: data
+      )
+    ).notTo(throwError())
+  }
+
+  // MARK: Tests
+
+  func test_build_givenInvalidConfiguration_shouldThrow() throws {
+    // given
+    let config = ApolloCodegenConfiguration(basePath: "not_a_path")
+
+    // then
+    expect(try ApolloCodegen.build(with: config)).to(throwError())
+  }
+
+  /// ```
+  /// SCHEMA:
+  /// type Query {
+  ///   books: [Book!]!
+  ///   authors: [Author!]!
+  /// }
+  ///
+  /// type Book {
+  ///   title: String!
+  ///   author: Author!
+  /// }
+  ///
+  /// type Author {
+  ///   name: String!
+  ///   books: [Book!]!
+  /// }
+  ///
+  /// OPERATION:
+  /// query getBooks {
+  ///   books {
+  ///     title
+  ///     name (this will cause the validation error)
+  ///   }
+  /// }
+  /// ```
+  func test_compileResults_givenOperation_withGraphQLErrors_shouldThrow() throws {
+    // given
+    let schemaPath = directoryURL.appendingPathComponent("schema.graphqls").path
+    createSchema(at: schemaPath)
+
+    let searchPath = directoryURL.appendingPathComponent("*.graphql").path
+    let data: Data =
+      """
+      query getBooks {
+        books {
+          title
+          name
+        }
+      }
+      """.data(using: .utf8)!
+    createOperationFile(for: data, inDirectory: directoryURL)
+
+    let config = ApolloCodegenConfiguration.FileInput(
+      schemaPath: schemaPath,
+      searchPaths: [searchPath]
```

Comment:
Looking at this now, I think it would be clearer if we named this `operationSearchPaths`.

One thing I love about tests is it forces you to consume the API you are building so you can see how clear the usage of it is. haha :)

Reply:
It's not only operation files being consumed here though. I need to add a test where `searchPaths` includes both a schema extension and operation to ensure that they're both being merged into the compilation result. I'm not 100% certain on the schema extension syntax though, will figure that out tomorrow.

Reply:
I'm struggling to get this test written, so I've created https://github.com/apollographql/apollo-ios/issues/2058 to keep track of it and I'll sync up with @IvanGoncharov to figure out where I'm going wrong in using graphql-js.